File: backend/native_engine/src/geo.c
#include "resq_native.h"

#include <math.h>

#define EARTH_RADIUS_KM 6371.0

static double deg2rad(double deg) {
    return deg * (M_PI / 180.0);
}

double resq_haversine_km(double lat1, double lng1, double lat2, double lng2) {
    double phi1 = deg2rad(lat1);
    double phi2 = deg2rad(lat2);
    double d_phi = deg2rad(lat2 - lat1);
    double d_lambda = deg2rad(lng2 - lng1);

    double a = sin(d_phi / 2.0) * sin(d_phi / 2.0)
             + cos(phi1) * cos(phi2) * sin(d_lambda / 2.0) * sin(d_lambda / 2.0);
    double c = 2.0 * asin(sqrt(a));
    return EARTH_RADIUS_KM * c;
}

int resq_is_within_radius(double lat1, double lng1, double lat2, double lng2, double radius_km) {
    return resq_haversine_km(lat1, lng1, lat2, lng2) <= radius_km ? 1 : 0;
}

void resq_radius_filter_batch(
    double origin_lat, double origin_lng, double radius_km,
    const double *lats, const double *lngs, size_t n,
    uint8_t *out_mask
) {
    for (size_t i = 0; i < n; i++) {
        out_mask[i] = resq_is_within_radius(origin_lat, origin_lng, lats[i], lngs[i], radius_km) ? 1 : 0;
    }
}

int resq_point_in_polygon(
    double lat, double lng,
    const double *polygon_lats, const double *polygon_lngs, size_t n
) {
    if (n < 3) {
        return 0;
    }

    int inside = 0;
    for (size_t i = 0, j = n - 1; i < n; j = i++) {
        double lat_i = polygon_lats[i], lng_i = polygon_lngs[i];
        double lat_j = polygon_lats[j], lng_j = polygon_lngs[j];

        int straddles = ((lat_i > lat) != (lat_j > lat));
        if (straddles) {
            double lng_intersect = lng_i + (lat - lat_i) * (lng_j - lng_i) / (lat_j - lat_i);
            if (lng < lng_intersect) {
                inside = !inside;
            }
        }
    }
    return inside;
}

File: backend/apps/common/geo.py
import math

EARTH_RADIUS_KM = 6371.0


def haversine_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Great-circle distance between two points in kilometres."""
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    d_phi = math.radians(lat2 - lat1)
    d_lambda = math.radians(lng2 - lng1)

    a = math.sin(d_phi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(d_lambda / 2) ** 2
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(a))


def filter_within_radius(queryset, lat: float, lng: float, radius_km: float):
    """Bounding-box pre-filter (cheap, indexable) then exact haversine cutoff (precise)."""
    from apps.common.models import GeoLocationMixin

    lat_min, lat_max, lng_min, lng_max = GeoLocationMixin.bounding_box(lat, lng, radius_km)
    candidates = queryset.filter(
        latitude__gte=lat_min,
        latitude__lte=lat_max,
        longitude__gte=lng_min,
        longitude__lte=lng_max,
    )
    return [
        obj
        for obj in candidates
        if obj.latitude is not None
        and obj.longitude is not None
        and haversine_km(lat, lng, float(obj.latitude), float(obj.longitude)) <= radius_km
    ]
